auth/context.go

```diff
@@ -0,0 +1,65 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import "context"
+
+var (
+	rawKey     = rawType{}
+	subjectKey = subjectType{}
+	groupsKey  = groupsType{}
+)
+
+type rawType struct{}
+type subjectType struct{}
```

Comment:
Why do we need this in a generic auth/context.go at all? These appear to be specific context data for example by oidc auth extension. Why not declare there on oidc implementation?
I think we should aim for every auth extension to declare its own context data.

Reply:
These are not specific to the OIDC auth extension. Those three pieces should be common to most authenticators, as they are common parts of any authorization system. For example, we might have a basic auth authenticator, a la Apache httpd: in such case, we would have a `.htpasswd` that would serve as the user data source (subject), and an optional `.htgroups`, linking groups with users like Unix' `/etc/group`. The raw field would be the base64 encoded `username:password` received by the HTTP server from the client.

> I think we should aim for every auth extension to declare its own context data.

I think every extension (not only auth) should be allowed to enhance the context data as they wish so that other components can make use of that information, but we should have a minimal set of options that are common to all authenticators, and this auth/context represents that in my opinion.

Reply:
Sorry, I still do not understand why this needs to be in the `auth` package and not in the specific extension's package which needs this. We are not expecting that we can mix auth extensions in any way, right? So why not move these declarations to the extension which plans to inject these values in the context?
I may be missing something but I think different authenticator extensions will need different values. Some authenticators will only need raw auth token others may need a subject or whatever it is called.
I don't understand what do we gain by making these concepts part of the public API of core and not just delegate it to individual auth extensions.

Reply:
>We are not expecting that we can mix auth extensions in any way, right?

We certainly expect auth extensions to provide data that will be consumed by other components in the pipeline. For instance, the routing processor can make decisions based on the principal or group membership (tenant "acme" should go to this Jaeger, tenant "ecorp" should go to that Zipkin, ...).

All auth extensions are expected to have at least the notion of principal (like a username, or a service account name, or ...) Raw can certainly be removed from the generic package, though. I would also keep the group information, as it's also a common aspect of auth no matter the backend used.

> I don't understand what do we gain by making these concepts part of the public API of core and not just delegate it to individual auth extensions.

Other components, like the routing processor I mentioned earlier, might provide tighter integration with the auth features, without requiring specific authenticators to exist.

Reply:
@jpkrohling can you please show how a routing processor could use the auth data? A sample code somewhere would be great to see and would help review this PR.

Reply:
It cannot right now because of this missing piece, but this is how a configuration would look like for the existing [routingprocessor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/routingprocessor):

```yaml
receivers:
  nop:

processors:
  routing:
    default_exporters:
    - otlp
    from_auth: membership
    table:
    - value: acme
      exporters: 
      - jaeger/acme
      - otlp/acme
    - value: globex
      exporters:
      - otlp/globex

exporters:
  otlp:
  otlp/acme:
  otlp/globex:
  jaeger/acme:
    endpoint: localhost:14250

service:
  pipelines:
    traces:
      receivers:
      - nop
      processors:
      - routing
      exporters:
      - jaeger/acme
      - otlp/acme
      - otlp/globex
```

The routingprocessor would probably need to group the data by auth/membership, but minimal changes would be required to the current code, probably isolated to this function here: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/36a7da6a4c29ddf81d118ec159e0a7a1682c6f21/processor/routingprocessor/routing.go#L133-L134 . Instead of extracting the route destination from the context, it would extract from the pdata.

Reply:
Looking at my notes, I think we did discuss something else though:

- the authenticators would place auth data into the context
- something at the junction point between receivers and processors would place this into pdata
- processors/exporters should rely on the data from pdata instead of the context

Other than that, the general idea remains.